versioned/tiered/rocksdb/src/main/java/org/projectnessie/versioned/rocksdb/RocksDBValueVisitor.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.rocksdb;
+
+import java.security.InvalidParameterException;
+import java.util.List;
+
+import org.projectnessie.versioned.impl.condition.ExpressionFunction;
+import org.projectnessie.versioned.impl.condition.ExpressionPath;
+import org.projectnessie.versioned.impl.condition.Value;
+import org.projectnessie.versioned.impl.condition.ValueVisitor;
+import org.projectnessie.versioned.store.Entity;
+
+/**
+ * This provides a separation of queries on @{ExpressionFunction} from the object itself.
+ * This uses the Visitor design pattern to retrieve object attributes.
+ */
+class RocksDBValueVisitor implements ValueVisitor<Function> {
+  private static class RocksDBExpressionPathVisitor implements ValueVisitor<ExpressionPath> {
+    @Override
+    public ExpressionPath visit(Entity entity) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ExpressionPath visit(ExpressionFunction value) {
+      final ExpressionFunction.FunctionName name = value.getName();
+      throw new UnsupportedOperationException(String.format("%s is not a supported top-level RocksDB function.", name));
+    }
+
+    @Override
+    public ExpressionPath visit(ExpressionPath value) {
+      return value;
+    }
+  }
+
+  private static final RocksDBExpressionPathVisitor EXPRESSION_PATH_VALUE_VISITOR = new RocksDBExpressionPathVisitor();
+
+  @Override
+  public Function visit(Entity entity) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public Function visit(ExpressionFunction value) {
```

Comment:
I bet we'll get comments on the fact that the visitors really only have a single case for both. It does make me wonder if we should use the visitors at all, since you already know what you're dealing with?

Reply:
I think it is worth waiting then refactoring if there are comments, and there will be comments.